src/power_grid_model/data_types.py

```diff
@@ -92,39 +92,50 @@
     - The first element and last element will therefore be 0 and the size of the data, respectively.
 """
 
-SparseBatchArray = dict[SparseDataComponentType, IndexPointer | SingleArray]
-"""
-A sparse batch array is a dictionary containing the keys `indptr` and `data`.
 
-- data: a :class:`SingleArray`. The exact dtype depends on the type of component.
-- indptr: an :class:`IndexPointer` representing the start and end indices for each batch scenario.
+class SparseBatchArray(TypedDict):
```

Comment:
😮 didn't know this existed. good find!

Reply:
Yeah on the retrospect we could have avoided the TypeVar thing.
We colud have had:

class ComponentType(total=False):
    node: SingleArray
    line: SingleArray
    ...

The totality makes these members optional.

Reply:
but that would replace dictionary datasets entirely, right?

Reply:
Oh, right. 
```
class Dataset(total=False):
    node: ComponentData
    line: ComponentData
    ...
```
I see now that we can't improve anything with this with the current implementation.